core/src/main/java/org/infinispan/interceptors/BaseAsyncInterceptor.java

```diff
@@ -282,6 +289,69 @@ public static InvocationStage makeStage(Object rv) {
       }
    }
 
+   /**
+    * Returns an InvocationStage if the provided CompletionStage is null, not completed or completed via exception.
+    * If these are not true the sync value is returned directly.
+    * @param stage wait for completion of this if not null
+    * @param syncValue sync value to return if stage is complete or as stage value
+    * @return invocation stage or sync value
+    */
+   public static Object stageOrSyncValue(CompletionStage<?> stage, Object syncValue) {
+      if (stage != null) {
+         CompletableFuture<?> future = stage.toCompletableFuture();
+         if (!future.isDone()) {
+            return asyncValue(stage.thenApply(v -> syncValue));
+         }
+         if (future.isCompletedExceptionally()) {
+            return asyncValue(stage);
+         }
+      }
+      return syncValue;
+   }
+
+   /**
+    * This method should be used instead of {@link #stageOrSyncValue(CompletionStage, Object)} when a
+    * {@link InvocationFinallyFunction} is used to properly handle the exception if any is present.
+    * @param stage
+    * @param syncValue
+    * @param throwable
+    * @return
+    */
+   public static Object stageOrSyncValue(CompletionStage<?> stage, Object syncValue, Throwable throwable) {
+      if (throwable == null) {
+         return stageOrSyncValue(stage, syncValue);
+      }
+      if (stage != null) {
+         CompletableFuture<?> future = stage.toCompletableFuture();
+         if (!future.isDone() || future.isCompletedExceptionally()) {
+            return asyncValue(
+                  stage.handle((ignore, t) -> {
+                     if (t != null) {
+                        throwable.addSuppressed(t);
+                     }
+                     return null;
+                  }).thenApply(ignore -> CompletableFutures.completedExceptionFuture(throwable))
```

Comment:
Hrmm I think I meant to make this `thenCompose` not then `thenApply`. I was worried about `CompletableFuture` wrapping the throwable in a CompletionException, without the compose.

To be honest I wasn't too worried about perf for this since it is an exception case (regarding the 2 lambdas).

Reply:
And you mean it is cheaper since they are store in the queue, correct? We also have to allocate first a `SimpleAsyncInvocationStage` as well. I wonder if we should have a way to eagerly create the `QueueAsyncInvocationStage` that is more friendly?